modules/xmpp/ChatRoom.js

```diff
@@ -958,10 +958,29 @@ ChatRoom.prototype.onMute = function (iq) {
 
 /**
  * Leaves the room. Closes the jingle session.
+ * @returns {Promise} which is resolved if XMPPEvents.MUC_LEFT is received less
+ * than 5s after sending presence unavailable. Otherwise the promise is
+ * rejected.
  */
 ChatRoom.prototype.leave = function () {
     this._dispose();
-    this.doLeave();
+    return new Promise((resolve, reject) => {
+        let timeout = setTimeout(() => onMucLeft(true), 5000);
+        let eventEmitter = this.eventEmitter;
+        function onMucLeft(doReject = false) {
+            eventEmitter.removeListener(XMPPEvents.MUC_LEFT, onMucLeft);
+            clearTimeout(timeout);
+            if(doReject) {
+                // the timeout expired
+                reject(new Error("The timeout for the confirmation about " +
+                    "leaving the room expired."));
+            } else {
+                resolve();
+            }
+        }
+        eventEmitter.on(XMPPEvents.MUC_LEFT, onMucLeft);
```

Comment:
I think assigning to a local variable wouldn't make it more readable and when you are debugging or logging errors from it the function is going to be displayed as anonymous function.

I agree that defining the function before using it looks confusing but the function is using the timeout variable and event emitter which are defined just before the function. So I don't think that moving the function will make it less confusing.